File: 98. Validate Binary Search Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
// version1: traverse (recursion)
public:
    bool isValidBST(TreeNode* root) {
        TreeNode* pre = NULL;
        return valid(root, pre);
    }
    bool valid(TreeNode* root, TreeNode*& pre) {
        if(!root) return true;
        if(!valid(root -> left, pre)) return false;
        if(pre && pre -> val >= root -> val) return false;
        pre = root;
        if(!valid(root -> right, pre)) return false;
        return true;
    }
};



/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
// version2: traverse (iteration)
    bool isValidBST(TreeNode* root) {
        stack<TreeNode*> st;
        TreeNode* pre = nullptr;
        while(!st.empty() || root) {
            while(root) {
                st.push(root);
                root = root -> left;
            }
            if(!st.empty()) {
                root = st.top();
                if(pre && pre -> val >= root -> val) return false;
                pre = root;
                st.pop();
                root = root -> right;
            }
        }
        return true;
    }
};


/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
struct Result {
// version3: Divide & Conquer
    Result(bool v, TreeNode* _min = nullptr, TreeNode* _max = nullptr) : isBST(v), minNode(_min), maxNode(_max) {}
    TreeNode* minNode;
    TreeNode* maxNode;
    bool isBST;
};
class Solution {
public:
    bool isValidBST(TreeNode* root) {
        return validate(root).isBST;
    }
    Result validate(TreeNode* root) {
        if(!root) return Result(true);
        Result left = validate(root -> left);
        Result right = validate(root -> right);
        if(!left.isBST || !right.isBST) return Result(false);
        if(left.maxNode && left.maxNode -> val >= root-> val) return Result(false);
        if(right.minNode && right.minNode -> val <= root-> val) return Result(false);  
        TreeNode* min = (left.minNode) ? left.minNode : root;
        TreeNode* max = (right.maxNode) ? right.maxNode : root;     
        return Result(true, min, max);
    }
};

File: 98. Validate Binary Search Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def isValidBST(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        last = None
        def helper(root):
            nonlocal last
            if not root: return True
            if root.left and not helper(root.left): return False
            if last and last.val >= root.val: return False
            last = root
            if root.right and not helper(root.right): return False
            return True
        return helper(root)
        
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def isValidBST(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        def helper(root):
            cur, st, last = root, [], None
            while cur or st:
                while cur:
                    st.append(cur)
                    cur = cur.left
                cur = st.pop()
                if last and last.val >= cur.val:
                    return False
                last = cur
                cur = cur.right
            return True
        return helper(root)        
        
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def isValidBST(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        def helper(root):
            if not root: return (True, None, None)
            left = helper(root.left)
            right = helper(root.right)
            if not left[0] or not right[0]: return (False, None, None)
            if root.left and left[2].val >= root.val: return (False, None, None)
            if root.right and right[1].val <= root.val: return (False, None, None)
            maxT = right[2] if root.right else root
            minT = left[1] if root.left else root
            return (True, minT, maxT)
        return helper(root)[0]        
        
        